arch: sw_isr: store device info in the table and add funtions to access

ycsin opened the pull request:
Change the internal function to `get_parent_entry`, which returns the entire entry of table.

Store the parent interrupt controller device in the `irq_parent_offset` table, and added 2 helper functions to:

1. determine the parent interrupt controller based on the IRQ
2. determine the IRQ of the parent interrupt controller

Declare the `struct _irq_parent_entry` in the header and added so that it can be used to test the functions in testsuites.

> [!NOTE]  
> This change will increase the LUT size for build with multi-level enabled by `CONFIG_NUM_x_LEVEL_AGGREGATORS * sizeof(struct device)`

___

> [!NOTE]  
> This PR is spined off from #62907, to hopefully make it easier to review, and is required to fix #63465

> [!IMPORTANT]
> Blocking #62907

ycsin committed:
arch: sw_isr: store device info in the table and add funtions to access

Change the internal function to `get_parent_entry`, which
returns the entire entry of table.

Store the parent interrupt controller device in the
`irq_parent_offset` table, and added 2 helper functions to:

1. determine the parent interrupt controller based on the IRQ
2. determine the IRQ of the parent interrupt controller

Declare the `struct _irq_parent_entry` in the header and added
`-` suffix to the struct so that it can be used to test the
functions in testsuites.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
tests: interrupt: add testcase for functions in the sw_isr_table

Validate the following functions in the sw_isr_table:
- z_get_sw_isr_table_idx
- z_get_sw_isr_device_from_irq
- z_get_sw_isr_irq_from_device

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt commented on arch/common/sw_isr_common.c:
```diff
@@ -16,12 +16,30 @@
+/* Filter out interrupt controllers with `has-multi-instance` prop */
+#define _FILTER_INTC(node_id)                                                                      \
+	COND_CODE_1(DT_NODE_HAS_PROP(node_id, interrupt_controller),                               \
+		    (DT_PROP_OR(node_id, has_multi_instance, 0)), (0))
```
TL;DR - The one issue I have here is with `has_multi_instance`. DT is implicitly multi-instance. Instead of using `has_multi_instance`, I would suggest rather using something like `DT_NUM_INST_STATUS_OKAY()` which conveys multi-instance status without an additional DT property. The number of instances of a dt-compat is implicitly part of DT, so it's not really necessary to explicitly add this property that should always be true.

galak commented on dts/bindings/interrupt-controller/interrupt-controller.yaml:
```diff
@@ -12,3 +12,6 @@ properties:
     type: int
     required: true
     description: Number of items to expect in an interrupt specifier
+  "has-multi-instance":
```
What's the intent of this property?

galak reviewed (dismissed):
We should NOT add a property for `has-multi-instance` to the devicetree, but scan the tree in some way to get the count of instances to determine its greater than 1.

ycsin commented:
@cfriedt @galak the `has-multi-instance` prop was misleading, I was using it as a shortcut to get the controller that this implementation supports.

This implementation doesn't support `INTMUX`, i.e. [intc_rv32m1_intmux](https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/interrupt_controller/intc_rv32m1_intmux.c), a single instance of it has multiple interrupts as child:

https://github.com/zephyrproject-rtos/zephyr/blob/b4da11f92967256e6a451f519c739ec95ea6175d/dts/riscv/openisa/rv32m1.dtsi#L88-L177

I've updated the filter to grab the interrupt controller that is the child of `/soc` instead, to filter out the root interrupt controllers

jhedberg commented:
@galak could you take another look?

fabiobaltieri commented:
(changing the assignees since this does not change any bindings anymore)

nashif commented on include/zephyr/sw_isr_table.h:
```diff
@@ -87,6 +94,25 @@ extern struct z_shared_isr_table_entry z_shared_sw_isr_table[];
+ *
+ * @return corresponding interrupt controller device in _sw_isr_table
+ */
+const struct device *z_get_sw_isr_device_from_irq(unsigned int irq);
```
did not do any extensive review, but this caught my attention, we should avoid declaring such function in public headers and give the impression they are public APIs anyone can use. Please move those to where they are being used or putting them in some private header.

ycsin committed:
arch: common: irq: relocate internal functions into a private header

Relocate new and existing internal software-managed table
access functions from the public `sw_isr_table.h` into a
private header that should only be accessed internally.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin commented on arch/common/include/sw_isr_common.h:
```diff
@@ -0,0 +1,57 @@
+/*
```
I've moved the helper functions into an internal header, is this what you meant? @nashif

nashif replied:
yeah, I may have overreacted to the sight of z_ APIs, it seems this header is full of them already. Need to spend some time on this and see what the status of the header and if it needs to be in the public domain in the first place. thanks.

ycsin commented:
rebased

fkokosinski reviewed (commented):
LGTM in general, one minor question.

Note that I'm reviewing this from the perspective of https://github.com/zephyrproject-rtos/zephyr/pull/62907.

fkokosinski commented on arch/common/shared_irq.c:
```diff
@@ -4,6 +4,8 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
+#include "include/sw_isr_common.h"
```
> Do you mean targeting the `include` path directly and just do

Yup, exactly

ycsin replied:
I have no preference over having it relative or targetted with `target_include_directories`, do you prefer the latter? Is the latter a standard practice for such private headers?

cfriedt replied:
Yeah, it's not completely formalized yet (and it would be great if tooling could automatically sort headers), but this has been around for a while. Based on Google's C++ coding guidelines.

https://github.com/zephyrproject-rtos/zephyr/issues/41543

ycsin commented on arch/common/CMakeLists.txt:
```diff
@@ -2,6 +2,10 @@
 
 zephyr_library()
 
+target_include_directories(${ZEPHYR_CURRENT_LIBRARY} PRIVATE
```
I hope I'm doing it this correctly

ycsin commented on arch/common/CMakeLists.txt:
```diff
@@ -2,6 +2,8 @@
 
 zephyr_library()
 
+zephyr_library_include_directories(include)
```
@fkokosinski I've added this and the header path used in the source isn't relative anymore

ycsin commented:
Rebased for the fix in #64630 to remove the workaround

cfriedt commented on arch/common/sw_isr_common.c:
This file (and the CMakeLists.txt) need some serious cleanup. Please do that in this PR. Maybe as separate commits.

1. remove the `#ifdef CONFIG_DYNAMIC_INTERRUPTS` / `#endif` in this file (we are more sophisticated than that in Zephyr)
2. adjust `CMakeLists.txt` to use `zephyr_library_sources_ifdef(CONFIG_DYNAMIC_INTERRUPTS sw_isr_common.c)`
3. `#define CAT_2ND_LVL_LIST(i, base) ...` does not need to be surrounded by `#ifdef CONFIG_2ND_LEVEL_INTERRUPTS` / `#endif`
4. Same as above with `#define CAT_3RD_LVL_LIST(i, base) ...`
5. `get_parent_offset()` should be static (and get rid of the `#ifdef CONFIG_MULTI_LEVEL_INTERRUPTS` / `#endif`
6. Use more `if (IS_ENABLED(CONFIG_MULTI_LEVEL_INTERRUPTS)` or `if (! IS_ENABLED(CONFIG_MULTI_LEVEL_INTERRUPTS)`
7. Use more `if (IS_ENABLED(CONFIG_3RD_LEVEL_INTERRUPTS)` and `if (! IS_ENABLED(CONFIG_3RD_LEVEL_INTERRUPTS)`
...

cfriedt replied:
Just to be more consistent code structure it would be a bit nicer to see `if (IS_ENABLED(CONFIG_2ND_LEVEL_INTERRUPTS)` instead of `if (IS_ENABLED(CONFIG_MULTI_LEVEL_INTERRUPTS)`. They are basically a logical equivalence, but that should be encoded in Kconfig rather than being skipped in code.

ycsin replied:
Can't fully remove `CONFIG_3RD_LEVEL_INTERRUPTS` because some Kconfigs used by the 3rd level variable `_lvl3_irq_list` initialization isn't defined

ycsin committed:
arch: common: refactor multi-level IRQ code

Refactor multi-level IRQ related code from `sw_isr_common.c` to
`multilevel_irq.c` to simplify `sw_isr_common` & macrologies.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
arch: common: guard the compilation of sw_isr_common with CMakeLists.txt

Instead of using a macro guard to prevent functions in
`sw_isr_common.c` from getting compiled when
`CONFIG_DYNAMIC_INTERRUPTS` isn't enabled, do that in
`CMakeLists.txt` instead.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

ycsin committed:
irq: multilevel: reduce the use of macros

Convert `#if defined CONFIG_*` to use
`if (IS_ENABLED(CONFIG_*))` and removed some macro guards since
they are safe to be compile.

Signed-off-by: Yong Cong Sin <ycsin@meta.com>

cfriedt commented on arch/common/multilevel_irq.c:
```diff
@@ -0,0 +1,167 @@
+
+#ifdef CONFIG_3RD_LEVEL_INTERRUPTS
+
+const struct _irq_parent_entry _lvl3_irq_list[CONFIG_NUM_3RD_LEVEL_AGGREGATORS]
```
Does this need to be conditionally defined?

E.g. can we remove the ifdef above, use `if (IS_ENABLED())` elsewhere, and let the linker discard unused code and data?

ycsin replied:
yeah it fails to compile as the initialization used some undefined macros when 3rd level isn't enabled

fkokosinski reviewed (approved):
LGTM from the perspective of https://github.com/zephyrproject-rtos/zephyr/pull/62907.

carlescufi commented:
@dcpleung can you take a look?

dcpleung reviewed (approved):
TBH... we should probably start looking into DAG or tree-like hierarchy for multi-level interrupts so we can avoid making this even complicated in the future.

ycsin commented:
> TBH... we should probably start looking into DAG or tree-like hierarchy for multi-level interrupts so we can avoid making this even complicated in the future.

Where can I read more about this? Any existing implementation?

dcpleung commented:
Maybe the Linux kernel? The idea is (probably) simple, where each interrupt controller is a node in a tree. For multi-level interrupts, each node is a child of some other node (except the root node). So you can have arbitrary levels of interrupts, and each controller can trace back to the root controller. Though, this is an overhaul of interrupt handling and is going to be a huge project... for soul(s) brave enough to tackle it.

cfriedt commented:
> Maybe the Linux kernel?

Yep. Linux would suffice. Of course, it's mostly about the model. The model will also drastically simplify shared interrupts with a marginal increase in complexity. It's also more stateful.

Zephyr's multi-level model (maybe surprisingly) started out simple too, of course 🤪
